rclip/model.py

```diff
@@ -30,17 +30,22 @@ def compute_image_features(self, images: List[Image.Image]) -> np.ndarray:
 
     return image_features.cpu().numpy()
 
-  def compute_text_features(self, text: str) -> np.ndarray:
+  def compute_text_features(self, text: Union[str, List[str]]) -> np.ndarray:
     with torch.no_grad():
       text_encoded = self._model.encode_text(clip.tokenize(text).to(self._device))
       text_encoded /= text_encoded.norm(dim=-1, keepdim=True)
 
     return text_encoded.cpu().numpy()
 
-  def compute_similarities_to_text(self, item_features: np.ndarray, text: str) -> List[Tuple[float, int]]:
-    text_features = self.compute_text_features(text)
+  def compute_similarities_to_text(
+      self, item_features: np.ndarray,
+      positive_queries: List[str], negative_queries: List[str]) -> List[Tuple[float, int]]:
 
-    similarities = (text_features @ item_features.T).squeeze(0).tolist()
+    positive_features = self.compute_text_features(positive_queries)
+    negative_features = self.compute_text_features(negative_queries)
+    text_features = np.add.reduce(positive_features) - np.add.reduce(negative_features)
```

Comment:
This is good enough for now :+1: , but we may have done something like: `np.add.reduce(self.compute_text_features(all_queries) * [1,1,1,-1,-1])` to compute all of the features at once. The array of ones and negative ones should be pre-created.

Reply:
I thought about that - but I think the multiply is more expensive.   If this times it correctly: 

        import numpy as np
        import timeit
        
        npos = nneg = 1000
        pos = np.random.rand(npos,512)
        neg = np.random.rand(nneg,512)
        
        timeit.timeit(lambda: np.add.reduce(pos) - np.add.reduce(neg),number=1000)
        
        all = np.random.rand(npos + nneg,512)
        signs = np.array([1] * npos + [-1] * nneg)
        
        timeit.timeit(lambda: np.add.reduce(signs * all.T),number=1000)

I'm getting 

        >>> timeit.timeit(lambda: np.add.reduce(signs * all.T),number=1000)
        2.642763003001164
        >>> timeit.timeit(lambda: np.add.reduce(pos) - np.add.reduce(neg),number=1000)
        0.7372207119988161

I also don't find the array of signs as easy to read (but maybe that's just me).

Reply:
@ramayer, thank you for the benchmark! I agree that readibility suffers from the "sign array" approach. My main concern was with the performance difference of calling `compute_text_features` once vs. twice.